metldata/builtin_transformations/aggregate/func.py

```diff
@@ -0,0 +1,238 @@
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Transformation and aggregation functions for the aggregate module."""
+
+from abc import ABC, abstractmethod
+from collections import Counter
+from typing import Any, Iterable, Optional
+
+from metldata.builtin_transformations.aggregate.models import (
+    MinimalClass,
+    MinimalNamedSlot,
+)
+from metldata.transform.base import MetadataTransformationError
+
+_FUNCTION_REGISTRY = {}
+
+
+class AggregationFunction(ABC):
+    """An abstract class for aggregation transformation functions"""
+
+    @classmethod
+    @abstractmethod
+    def func(cls, data: Iterable[Any]) -> Any:
+        """Transforms input data."""
+
+    @classmethod
+    @abstractmethod
+    def result_range_name(cls) -> str:
+        """Returns the name of the range of the data produced by func."""
+
+    @classmethod
+    @abstractmethod
+    def result_range_cls_def(cls) -> Optional[MinimalClass]:
+        """Returns the class definition of the data produced by func. None if
+        the function result range is a type rather than a class."""
+
+    @classmethod
+    @abstractmethod
+    def result_multivalued(cls) -> bool:
+        """Returns whether or not the transformation function produces a single
+        value or a list."""
+
+
+def register_function(func: type[AggregationFunction]) -> type[AggregationFunction]:
+    """Registers a function in the aggregation function registry."""
+    _FUNCTION_REGISTRY[func.__name__] = func
+    return func
+
+
+class CopyAggregation(AggregationFunction, ABC):
+    """An abstract base class for type-specific copy aggregation functions."""
+
+    @classmethod
+    def _extract_single_value(cls, data: Iterable[Any]) -> Any:
+        iterator = iter(data)
+        value = next(iterator)
+        try:
+            next(iterator)
+        except StopIteration:
+            return value
+        raise MetadataTransformationError(
+            "Multiple values passed to copy aggregation function where only a"
+            " single value was expected."
+        )
+
+
+@register_function
+class StringListCopyAggregation(CopyAggregation):
+    """Transformation that returns a single"""
+
+    @classmethod
+    def func(cls, data: Iterable[Any]) -> list[str]:
+        return list[str](data)
+
+    @classmethod
+    def result_range_name(cls) -> str:
+        return "string"
```

Comment:
"string" comes from linkml:types

Reply:
For the records: "string" is in fact the correct LinkML root type. I was confused because it is displayed capitalized as "String" in the [LinkML docs](https://linkml.io/linkml-model/docs/String/) and because our own docs said it was a "defined" type. "str" ist the "base type" (underlying Python type).